base/utfconvert.jl

```diff
@@ -0,0 +1,444 @@
+# This file is a part of Julia. License is MIT: http://julialang.org/license
+
+# Functions to convert to different UTF encodings
+
+# Quickly copy and set trailing \0
+@inline function fast_utf_copy(T::Type{UInt16}, len, dat)
+    @inbounds return UTF16String(setindex!(copy!(Vector{T}(len), dat), 0, len))
+end
+@inline function fast_utf_copy(T::Type{Char}, len, dat)
+    @inbounds return UTF32String(setindex!(copy!(Vector{T}(len), dat), 0, len))
+end
+
+# Get rest of character ch from 3-byte UTF-8 sequence in dat
+@inline function get_utf8_3(dat, pos, ch)
+    @inbounds return ((ch & 0xf) << 12) | (UInt32(dat[pos-1] & 0x3f) << 6) | (dat[pos] & 0x3f)
+end
+
+# Get rest of character ch from 4-byte UTF-8 sequence in dat
+@inline function get_utf8_4(dat, pos, ch)
+    @inbounds return (((ch & 0x7) << 18)
+                        | (UInt32(dat[pos-2] & 0x3f) << 12)
+                        | (UInt32(dat[pos-1] & 0x3f) << 6)
+                        | (dat[pos] & 0x3f))
+end
+
+# Output a character as a 4-byte UTF-8 sequence
+@inline function output_utf8_4(buf, out, ch)
+    @inbounds begin
+        buf[out + 1] = 0xf0 | (ch >>> 18)
+        buf[out + 2] = 0x80 | ((ch >>> 12) & 0x3f)
+        buf[out + 3] = 0x80 | ((ch >>> 6) & 0x3f)
+        buf[out + 4] = 0x80 | (ch & 0x3f)
+    end
+end
+
+#=
+"""
+@brief      Converts an AbstractString to a UTF16String
```

Comment:
I don't see how `@doc` could happen before strings are defined, it needs to parse strings into Markdown. We may be stuck with a sep. file (as is trialed at the moment in `basedocs.jl`), I guess another option in principle is to do a second pass of the files and just apply the docstrings (don't run anything)...

I think the idea was (by 0.4 release) for the `@doc` macro to be optional you just write:
```
"""docstring *with markdown*"""
function f(...
```

But this is way off topic!